diff --git a/Source/WTF/wtf/FastMalloc.cpp b/Source/WTF/wtf/FastMalloc.cpp
index f560f56b02a8..30e43b3b0e8c 100644
--- a/Source/WTF/wtf/FastMalloc.cpp
+++ b/Source/WTF/wtf/FastMalloc.cpp
@@ -29,6 +29,10 @@
 #include <string.h>
 #include <wtf/CheckedArithmetic.h>
 
+#if OS(DARWIN)
+#include <malloc/malloc.h>
+#endif
+
 #if OS(WINDOWS)
 #include <windows.h>
 #else
diff --git a/Source/WTF/wtf/PlatformGTK.cmake b/Source/WTF/wtf/PlatformGTK.cmake
index bb01bfeeac63..b5b9de3a9e92 100644
--- a/Source/WTF/wtf/PlatformGTK.cmake
+++ b/Source/WTF/wtf/PlatformGTK.cmake
@@ -24,6 +24,7 @@ if (CMAKE_SYSTEM_NAME MATCHES "Linux")
     )
 elseif (CMAKE_SYSTEM_NAME MATCHES "Darwin")
     list(APPEND WTF_PUBLIC_HEADERS
+        spi/darwin/OSVariantSPI.h
         spi/darwin/ProcessMemoryFootprint.h
     )
 endif ()
diff --git a/Source/WTF/wtf/RandomDevice.cpp b/Source/WTF/wtf/RandomDevice.cpp
index db85f5062bff..1e35bcd6e0d0 100644
--- a/Source/WTF/wtf/RandomDevice.cpp
+++ b/Source/WTF/wtf/RandomDevice.cpp
@@ -41,6 +41,7 @@
 #endif
 
 #if OS(DARWIN)
+#include <CommonCrypto/CommonCryptor.h>
 #include <CommonCrypto/CommonRandom.h>
 #endif
 
diff --git a/Source/WTF/wtf/WTFConfig.cpp b/Source/WTF/wtf/WTFConfig.cpp
index 87b9421241d7..263ca5f4ac7f 100644
--- a/Source/WTF/wtf/WTFConfig.cpp
+++ b/Source/WTF/wtf/WTFConfig.cpp
@@ -30,7 +30,7 @@
 #include <wtf/Lock.h>
 #include <wtf/StdLibExtras.h>
 
-#if OS(DARWIN)
+#if PLATFORM(COCOA)
 #include <wtf/spi/cocoa/MachVMSPI.h>
 #include <mach/mach.h>
 #elif OS(LINUX)
@@ -62,7 +62,7 @@ namespace WTF {
 #if ENABLE(UNIFIED_AND_FREEZABLE_CONFIG_RECORD)
 void setPermissionsOfConfigPage()
 {
-#if OS(DARWIN)
+#if PLATFORM(COCOA)
     static std::once_flag onceFlag;
     std::call_once(onceFlag, [] {
         mach_vm_address_t addr = bitwise_cast<uintptr_t>(static_cast<void*>(WebConfig::g_config));
@@ -86,7 +86,7 @@ void setPermissionsOfConfigPage()
 
         RELEASE_ASSERT(result == KERN_SUCCESS);
     });
-#endif // OS(DARWIN)
+#endif // PLATFORM(COCOA)
 }
 #endif // ENABLE(UNIFIED_AND_FREEZABLE_CONFIG_RECORD)
 
@@ -107,7 +107,7 @@ void Config::permanentlyFreeze()
     int result = 0;
 
 #if ENABLE(UNIFIED_AND_FREEZABLE_CONFIG_RECORD)
-#if OS(DARWIN)
+#if PLATFORM(COCOA)
     enum {
         AllowPermissionChangesAfterThis = false,
         DisallowPermissionChangesAfterThis = true
diff --git a/Source/WTF/wtf/posix/OSAllocatorPOSIX.cpp b/Source/WTF/wtf/posix/OSAllocatorPOSIX.cpp
index 42bbf8123735..7510126db46e 100644
--- a/Source/WTF/wtf/posix/OSAllocatorPOSIX.cpp
+++ b/Source/WTF/wtf/posix/OSAllocatorPOSIX.cpp
@@ -44,7 +44,7 @@
 #endif // OS(DARWIN)
 #endif // ENABLE(JIT_CAGE)
 
-#if OS(DARWIN)
+#if PLATFORM(COCOA)
 #include <wtf/spi/cocoa/MachVMSPI.h>
 #endif
 
